path_oidc.go

```diff
@@ -0,0 +1,194 @@
+package jwtauth
+
+import (
+	"context"
+	"time"
+
+	oidc "github.com/coreos/go-oidc"
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/vault/helper/base62"
+	"github.com/hashicorp/vault/helper/strutil"
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+	"golang.org/x/oauth2"
+)
+
+// base62 char length for state and nonce parameters, ~160 bits entropy per rfc6749#section-10.10
+const stateLength = 27
+
+var oidcStateTimeout = 2 * time.Minute
+
+// oidcState is created when an authURL is requested. The state identifier is
+// passed throughout the OAuth process.
+type oidcState struct {
+	rolename    string
+	expiration  time.Time
+	nonce       string
+	redirectURI string
+}
+
+func pathOIDC(b *jwtAuthBackend) []*framework.Path {
+	return []*framework.Path{
+		{
+			Pattern: `oidc/auth_url`,
+			Fields: map[string]*framework.FieldSchema{
+				"role": {
+					Type:        framework.TypeLowerCaseString,
+					Description: "The role to issue an OIDC authorization URL against.",
+				},
+				"redirect_uri": {
+					Type:        framework.TypeString,
+					Description: "The OAuth redirect_uri to use in the authorization URL.",
+				},
+			},
+			Callbacks: map[logical.Operation]framework.OperationFunc{
+				logical.UpdateOperation: b.authURL,
+			},
+		},
+	}
+}
+
+// authURL returns a URL used for redirection to receive an authorization code.
+// This path requires a role name, or that a default_role has been configured.
+// Because this endpoint is unauthenticated, the response to invalid or non-OIDC
+// roles is intentionally non-descriptive and will simply be an empty string.
+//
+// TODO: I think we should probably just log all errors and return nil instead of
+//       the usual "return nil, err" pattern. We can still return a user error when
+//       they've not provided a required field though.
+func (b *jwtAuthBackend) authURL(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	authCodeURL := ""
+
+	config, err := b.config(ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+	if config == nil {
+		return logical.ErrorResponse("could not load configuration"), nil
+	}
+
+	roleName := d.Get("role").(string)
+	if roleName == "" {
+		roleName = config.DefaultRole
+		if roleName == "" {
+			return logical.ErrorResponse("missing role"), nil
+		}
+	}
+
+	role, err := b.role(ctx, req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+
+	redirectURI := d.Get("redirect_uri").(string)
+	if redirectURI == "" {
+		return logical.ErrorResponse("missing redirect_uri"), nil
+	}
+
+	if role != nil && role.RoleType == "oidc" {
```

Comment:
Do we want to return early if role type != oidc?

Reply:
Yeah that's a little simpler.